File: src/leetcode/2026/04/Solution819.cpp
#include <string>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <cctype>
#include <iostream>

using namespace std;

class Solution819 {
public:
    string mostCommonWord(string paragraph, vector<string>& banned) {
        unordered_set<string> bannedSet(banned.begin(), banned.end());
        unordered_map<string, int> wordCount;
        string word;
        for (char c : paragraph) {
            if (isalpha(c)) {
                word += tolower(c);
            } else if (!word.empty()) {
                if (bannedSet.find(word) == bannedSet.end()) {
                    wordCount[word]++;
                }
                word.clear();
            }
        }
        // Check the last word if it exists
        if (!word.empty() && bannedSet.find(word) == bannedSet.end()) {
            wordCount[word]++;
        }

        string mostCommon;
        int maxCount = 0;
        for (const auto& entry : wordCount) {
            if (entry.second > maxCount) {
                mostCommon = entry.first;
                maxCount = entry.second;
            }
        }
        return mostCommon;
    }
};

int main() {
    Solution819 solution;

    string paragraph = "Bob hit a ball, the hit BALL flew far after it was hit.";
    vector<string> banned = {"hit"};
    string result = solution.mostCommonWord(paragraph, banned);
    // Expected output: "ball"
    cout << result << endl;

    paragraph = "a.";
    banned = {};
    result = solution.mostCommonWord(paragraph, banned);
    // Expected output: "a"
    cout << result << endl;

    return 0;
}

File: src/leetcode/2026/04/Solution819.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class Solution819 {
    public String mostCommonWord(String paragraph, String[] banned) {
        String[] words = paragraph.toLowerCase().split("[^a-z]+");
        Set<String> bannedSet = new HashSet<>(Arrays.asList(banned));
        Map<String, Integer> wordCount = new HashMap<>();

        for (String word : words) {
            if (!bannedSet.contains(word)) {
                wordCount.put(word, wordCount.getOrDefault(word, 0) + 1);
            }
        }

        String mostCommon = "";
        int maxCount = 0;

        for (Map.Entry<String, Integer> entry : wordCount.entrySet()) {
            if (entry.getValue() > maxCount) {
                mostCommon = entry.getKey();
                maxCount = entry.getValue();
            }
        }

        return mostCommon;
    }

    public static void main(String[] args) {
        Solution819 solution = new Solution819();

        String paragraph1 = "Bob hit a ball, the hit BALL flew far after it was hit.";
        String[] banned1 = { "hit" };
        String result1 = solution.mostCommonWord(paragraph1, banned1);
        System.out.println(result1); // Output: "ball"

        String paragraph2 = "a.";
        String[] banned2 = {};
        String result2 = solution.mostCommonWord(paragraph2, banned2);
        System.out.println(result2); // Output: "a"
    }
}
